### ISSUE TITLE
inout returns non-positive numbers

### ISSUE THREAD

**Issue by gchrupala** [spokenlanguage/platalea#21]
https://github.com/spokenlanguage/platalea/blob/510d5ef9270411dc971c19d988c30bed4c1c20d6/platalea/encoders.py#L443
For some inputs this returns negative numbers, which causes failures elsewhere and also logically makes no sense.

**Comment by egpbos** [comment]
There is an ad-hoc fix currently in `encoders.py` in `SpeechEncoderBottom.forward`:

```py
length = inout(self.Conv, length)
# FIXME. Avoid negative lengths which `inout` returns in some cases.
length = length.clamp(min=1)
```

Can we just add that `clamp` to the returned value of `inout`? Or are there more complex cases we need to take into account?

**Comment by cwmeijer** [comment]
There are valid cases where the output is smaller than 1. I guess the only possible such value is 0. This happens when the (dilated) kernel is greater than the (padded) input.

I think we should add a few unit tests for this one (regression tests) as this is no trivial task, the function is performing. 
so to fix:
- [ ] add at least one test for the negative outcome (small input and padding, large kernel and dilation)
- [ ] probably add a min/clamp to make it return at least 0.

Also, added my calculations because it has nice colours. I'll have a look if what is in the function right now, is the same as what I ended up with, otherwise I will discuss it first with your guys before correcting it. My calculation also allows for negative outcomes, so it is also missing a min(..., 0).
![image](https://user-images.githubusercontent.com/6087314/104573782-995ab680-5655-11eb-9493-b83887c9f7b1.png)

**Comment by bhigy** [comment]
I double checked the formula we currently have and the one in [pytorch's documentation for the conv1d](https://pytorch.org/docs/master/generated/torch.nn.Conv1d.html#conv1d), everything seems fine to me. Based on the hyperparameters we typically use, there could be indeed be a case where L<sub>out</sub> ends up negative when L<sub>in</sub> < 4. That might be what happened to Grzegorz, in which case clamping is probably a good solution (it shouldn't hurt in any case).
